returnn/tf/layers/basic.py

```diff
@@ -2685,6 +2685,83 @@ def get_out_data_from_opts(cls, name, axis, dim=1, sources=(), **kwargs):
     return data
 
 
+class RepeatLayer(_ConcatInputLayer):
+  """
+  A wrapper around tf.repeat, but supports an additional batch axis for the durations
+  The sum of the repetitions has to be non-zero for each sequence in the batch.
+
+  This layer can only be used with Tensorflow 1.15.0 or newer.
+  """
+  layer_class = "repeat"
+
+  def __init__(self, repetitions, axis="T", **kwargs):
+    """
+    :param LayerBase repetitions: number of repetitions for each sequence and position in target axis, [B,T] or [T,B]
+    :param str|int axis: (dynamic) axis for repetition (currently only time axis is supported)
+    """
+    super(RepeatLayer, self).__init__(**kwargs)
+    repetitions_data = repetitions.output
+    input_data = self.input_data
+    assert self.input_data
+    original_axis = input_data.get_axis_from_description(axis)
+    if original_axis < input_data.batch_dim_axis:
+      input_data = input_data.copy_as_batch_major().copy_move_axis(original_axis+1, 1)
+    else:
+      input_data = input_data.copy_as_batch_major().copy_move_axis(original_axis, 1)
+    repetitions_data = repetitions_data.copy_as_batch_major()
+    # pad the target axis
+    paddings = [(0, 1) if i == 1 else (0, 0) for i in range(input_data.batch_ndim)]
+    padded_data = tf.pad(input_data.placeholder, paddings)  # [B, T+1, ...]
+    # those are the sequence lengths after expansion
+    target_seq_len = tf.reduce_sum(repetitions_data.placeholder, axis=1)  # [B], the new size_placeholder for 'T'
+    # maximum sequence length after expansion
+    max_duration = tf.reduce_max(target_seq_len)  # [] == T'
+    # the new padding is the difference of the maximum to the new target
+    target_padding_steps = tf.expand_dims(max_duration - target_seq_len, 1)  # [B, 1]
+    # add the repetitions for the artificial padding position
+    target_repetitions = tf.concat([repetitions_data.placeholder, target_padding_steps], axis=1)  # [B, T+1]
```

Comment:
Better use `tf_util.concat_with_opt_broadcast` instead of `tf.concat`.

Reply:
What would be the reason?
```
    target_repetitions = tf_util.concat_with_opt_broadcast(
      [repetitions_data.placeholder, target_padding_steps], allow_broadcast=[False, False], axis=1)  # [B, T+1
```
looks more complicated, and does exactly the same.

Reply:
Two reasons:
* It makes the broadcasting explicit. (Explicit is often better than implicit.)
* This also works on earlier TF versions. (I don't exactly remember since when they support broadcasting in concat.)

Well, maybe both are not so important. Your choice.

Reply:
As `tf_util.concat_with_opt_broadcast` will return exactly the same `tf.concat` command as before, there is no functional difference, so I would keep it as it is.

Reply:
It's not about that this is the same (in newer TF versions). It's about:

* Having broadcasting explicit. (Explicit is often better than implicit.)
* This also works on earlier TF versions.

Reply:
So you just want to have it explicitly stated that there is no broadcasting allowed, in case the code is changed later?
I am not sure why you are mentioning earlier TF versions. This will work in any TF version.

Also TF might do broadcasting on its own, even if if set to false, if I understand you correctly. There is no check for that in `concat_with_opt_broadcast`, it just returns `tf.concat`:
```
  if all([not a for a in allow_broadcast]):
    return tf.concat(values, axis=axis)
```
So using the `tf_util` functions makes it even less explicit, as you can read it as "no broadcasting will be done in any way".

Reply:
> So you just want to have it explicitly stated that there is no broadcasting allowed, in case the code is changed later?

No. I said it is often better to explicitly state that sth (here: broadcasting) is done/happening. Otherwise it would be implicit.

> I am not sure why you are mentioning earlier TF versions. This will work in any TF version.

No. As I said, earlier TF versions do not support broadcasting in `tf.concat`.

> Also TF might do broadcasting on its own, even if if set to false, if I understand you correctly. There is no check for that in `concat_with_opt_broadcast`, it just returns `tf.concat`:
> 
> ```
>   if all([not a for a in allow_broadcast]):
>     return tf.concat(values, axis=axis)
> ```

This is because `tf.concat` did not support broadcasting earlier.

But anyway, `allow_broadcasting` is the other way around: It is not about disallowing broadcasting. It is about allowing it (explicitly).

You could extend the code now for newer TF versions, if you want to also check the other other way around (i.e. if allow=False, it is explicitly disallowed).

> So using the `tf_util` functions makes it even less explicit, as you can read it as "no broadcasting will be done in any way".

No. It is more explicit, because you explicitly write that you want to have broadcasting.

Reply:
I do not understand your comments. Here is no broadcasting happening.

If there is broadcasting, it definitely makes sense to write it explicit, and I understand that older TF version do not support this.

But in this case there is no broadcasting, I would rather disallow broadcasting here explicitly, as any broadcasting of the repetitions can only lead to incorrect results, but will never be correct.

Reply:
Ah sorry, I somehow thought there would be broadcasting happening here. (Probably because of the "1" and I did not read it carefully enough.)
Then you can ignore everything what I wrote. `tf.concat` is the right choice.